File: src/fusion/features/crowd_beliefs.py
"""
ZINC-FUSION-V15: Crowd Beliefs Feature Engineering

Extracts behavioral signals from Polymarket prediction markets for specialist training.
Forward-looking complement to backward-looking EPU indices.

Academic Foundation:
- Prediction markets aggregate dispersed information efficiently (NBER/Brookings)
- Favourite/longshot bias: high-likelihood events underpriced
- Probability momentum more predictive than levels (Wiley commodity sentiment study)
- 10-month lookback window optimal for sentiment signals

Cross-Specialist Routing:
- trump_effect: trump, executive, doge, deportation
- china: china, taiwan, trade_war
- tariff: tariff, import, trade
- biofuel: rfs, ethanol, epa, mandate
- energy: oil, sanctions, opec
- fed: fed, rates, inflation, recession
- volatility: vix, crash, crisis

Usage:
    from src.fusion.features.crowd_beliefs import CrowdBeliefsFeatureEngine

    engine = CrowdBeliefsFeatureEngine(conn)
    features = engine.compute_features(
        specialist='trump_effect',
        start_date='2025-01-01',
        end_date='2026-01-10'
    )
"""

import logging
from datetime import date, datetime, timedelta
from typing import Dict, List, Optional, Any
from dataclasses import dataclass

import numpy as np
import pandas as pd
import psycopg2
from psycopg2.extras import RealDictCursor

logger = logging.getLogger(__name__)


# =============================================================================
# CONFIGURATION
# =============================================================================

# Event categories mapped to specialists
CATEGORY_SPECIALIST_MAP = {
    'tariff': ['tariff', 'trump_effect', 'china'],
    'china': ['china', 'trump_effect'],
    'taiwan': ['china', 'volatility'],
    'trump': ['trump_effect'],
    'doge': ['trump_effect', 'fed'],
    'executive': ['trump_effect'],
    'deportation': ['trump_effect'],
    'immigration': ['trump_effect'],
    'fed': ['fed', 'volatility'],
    'rates': ['fed'],
    'inflation': ['fed', 'energy'],
    'recession': ['fed', 'volatility'],
    'biofuel': ['biofuel'],
    'energy': ['energy'],
    'volatility': ['volatility'],
    'trade': ['tariff', 'china'],
    'deficit': ['fed', 'trump_effect'],
}

# Key event slugs to track (high signal-to-noise)
KEY_EVENT_SLUGS = {
    'trump_effect': [
        'tariff-revenue',
        'trump-deport',
        'doge-cut',
        'trump-executive',
    ],
    'china': [
        'china-taiwan',
        'china-tariff',
        'china-trade',
    ],
    'tariff': [
        'tariff-revenue',
        'tariff-rate',
    ],
    'biofuel': [
        'rfs-mandate',
        'ethanol-policy',
        'epa-waiver',
    ],
    'fed': [
        'fed-rate',
        'recession',
        'inflation-target',
    ],
    'volatility': [
        'market-crash',
        'vix-spike',
        'recession',
    ],
}


@dataclass
class CrowdBeliefSnapshot:
    """Single point-in-time crowd belief state."""
    as_of_date: date
    event_slug: str
    outcome_question: str
    implied_prob_yes: float
    attention_index_24h: float
    attention_index_7d: float
    prob_momentum_24h: Optional[float]
    prob_momentum_7d: Optional[float]
    consensus_strength: float
    days_to_resolution: Optional[int]
    specialist_tags: List[str]


# =============================================================================
# FEATURE ENGINE
# =============================================================================

class CrowdBeliefsFeatureEngine:
    """
    Extracts behavioral features from Polymarket crowd beliefs.

    Features extracted per specialist:
    - Probability levels (current belief state)
    - Momentum signals (rate of change in beliefs)
    - Attention signals (betting activity spikes)
    - Consensus signals (crowd agreement level)
    - Time decay (urgency as resolution approaches)

    Academic notes:
    - Momentum more predictive than levels (Wiley study)
    - Attention spikes indicate news/events
    - Calibration degrades far from resolution date
    """

    def __init__(self, conn: psycopg2.extensions.connection):
        """
        Initialize with PostgreSQL connection.

        Args:
            conn: psycopg2 connection to fusion database
        """
        self.conn = conn

    def _load_beliefs(
        self,
        specialist: str,
        start_date: date,
        end_date: date
    ) -> pd.DataFrame:
        """
        Load crowd beliefs tagged to a specific specialist.

        Args:
            specialist: Specialist bucket name (e.g., 'trump_effect')
            start_date: Start of date range
            end_date: End of date range

        Returns:
            DataFrame with raw belief data
        """
        query = """
        SELECT
            DATE(captured_at) AS as_of_date,
            event_slug,
            outcome_question,
            implied_prob_yes,
            implied_prob_no,
            attention_index_24h,
            attention_index_7d,
            prob_momentum_24h,
            prob_momentum_7d,
            consensus_strength,
            event_category,
            specialist_tags,
            days_to_resolution,
            event_resolution_date,
            raw_betting_volume_usd,
            raw_liquidity_usd
        FROM raw.crowd_beliefs_event
        WHERE %s = ANY(specialist_tags)
          AND DATE(captured_at) BETWEEN %s AND %s
        ORDER BY captured_at
        """

        with self.conn.cursor(cursor_factory=RealDictCursor) as cur:
            cur.execute(query, (specialist, start_date, end_date))
            rows = cur.fetchall()

        if not rows:
            logger.warning(f"No crowd beliefs found for {specialist} in date range")
            return pd.DataFrame()

        return pd.DataFrame(rows)

    def _compute_probability_features(
        self,
        df: pd.DataFrame,
        specialist: str
    ) -> pd.DataFrame:
        """
        Compute probability-based features.

        Features:
        - prob_max_yes: Highest belief for negative outcome
        - prob_avg_yes: Average belief across tracked events
        - prob_weighted_avg: Volume-weighted average belief
        """
        if df.empty:
            return pd.DataFrame()

        features = df.groupby('as_of_date').agg({
            'implied_prob_yes': ['max', 'mean', 'std'],
            'implied_prob_no': ['max', 'mean'],
        }).reset_index()

        # Flatten column names
        features.columns = [
            'as_of_date',
            f'{specialist}_prob_max_yes',
            f'{specialist}_prob_avg_yes',
            f'{specialist}_prob_std_yes',
            f'{specialist}_prob_max_no',
            f'{specialist}_prob_avg_no',
        ]

        return features

    def _compute_momentum_features(
        self,
        df: pd.DataFrame,
        specialist: str
    ) -> pd.DataFrame:
        """
        Compute momentum-based features (rate of change in beliefs).

        Research: Momentum more predictive than levels (Wiley study)

        Features:
        - momentum_24h_max: Largest 24h shift
        - momentum_7d_avg: Average 7d shift
        - momentum_direction: Net directional shift
        """
        if df.empty:
            return pd.DataFrame()

        # Filter out nulls for momentum
        mom_df = df.dropna(subset=['prob_momentum_24h', 'prob_momentum_7d'])

        if mom_df.empty:
            return pd.DataFrame()

        features = mom_df.groupby('as_of_date').agg({
            'prob_momentum_24h': ['max', 'min', 'mean'],
            'prob_momentum_7d': ['max', 'min', 'mean'],
        }).reset_index()

        features.columns = [
            'as_of_date',
            f'{specialist}_momentum_24h_max',
            f'{specialist}_momentum_24h_min',
            f'{specialist}_momentum_24h_avg',
            f'{specialist}_momentum_7d_max',
            f'{specialist}_momentum_7d_min',
            f'{specialist}_momentum_7d_avg',
        ]

        # Add directional signal
        features[f'{specialist}_momentum_direction'] = np.sign(
            features[f'{specialist}_momentum_7d_avg']
        )

        return features

    def _compute_attention_features(
        self,
        df: pd.DataFrame,
        specialist: str
    ) -> pd.DataFrame:
        """
        Compute attention-based features (event detection).

        Research: Attention spikes indicate news/events

        Features:
        - attention_spike: Max attention across events
        - attention_avg: Average attention level
        - attention_divergence: Attention change without prob change
        """
        if df.empty:
            return pd.DataFrame()

        features = df.groupby('as_of_date').agg({
            'attention_index_24h': ['max', 'mean'],
            'attention_index_7d': ['max', 'mean'],
        }).reset_index()

        features.columns = [
            'as_of_date',
            f'{specialist}_attention_spike_24h',
            f'{specialist}_attention_avg_24h',
            f'{specialist}_attention_spike_7d',
            f'{specialist}_attention_avg_7d',
        ]

        # Attention divergence: high attention but low momentum = noise
        # (This would need momentum data, so compute it separately)

        return features

    def _compute_consensus_features(
        self,
        df: pd.DataFrame,
        specialist: str
    ) -> pd.DataFrame:
        """
        Compute consensus-based features (crowd uncertainty).

        Features:
        - consensus_avg: Average consensus strength
        - consensus_min: Minimum (most uncertain) consensus
        - crowd_uncertainty: 1 - consensus (higher = more uncertain)
        """
        if df.empty:
            return pd.DataFrame()

        features = df.groupby('as_of_date').agg({
            'consensus_strength': ['mean', 'min', 'max'],
        }).reset_index()

        features.columns = [
            'as_of_date',
            f'{specialist}_consensus_avg',
            f'{specialist}_consensus_min',
            f'{specialist}_consensus_max',
        ]

        # Crowd uncertainty is inverse of consensus
        features[f'{specialist}_crowd_uncertainty'] = (
            1 - features[f'{specialist}_consensus_avg']
        )

        return features

    def _compute_urgency_features(
        self,
        df: pd.DataFrame,
        specialist: str
    ) -> pd.DataFrame:
        """
        Compute time-decay / urgency features.

        Research: Calibration degrades far from resolution date

        Features:
        - resolution_urgency: 1/sqrt(days) - more weight as event approaches
        - near_term_events: Count of events resolving within 30 days
        """
        if df.empty:
            return pd.DataFrame()

        # Filter for events with resolution dates
        res_df = df.dropna(subset=['days_to_resolution'])

        if res_df.empty:
            return pd.DataFrame()

        # Add urgency score
        res_df = res_df.copy()
        res_df['urgency'] = 1 / np.sqrt(res_df['days_to_resolution'].clip(lower=1))

        features = res_df.groupby('as_of_date').agg({
            'urgency': ['max', 'mean'],
            'days_to_resolution': [
                'min',  # Nearest resolution
                lambda x: (x <= 30).sum(),  # Count near-term
            ],
        }).reset_index()

        features.columns = [
            'as_of_date',
            f'{specialist}_urgency_max',
            f'{specialist}_urgency_avg',
            f'{specialist}_nearest_resolution_days',
            f'{specialist}_near_term_event_count',
        ]

        return features

    def compute_features(
        self,
        specialist: str,
        start_date: date,
        end_date: date
    ) -> pd.DataFrame:
        """
        Compute all crowd belief features for a specialist.

        Args:
            specialist: Specialist bucket name (e.g., 'trump_effect')
            start_date: Start of date range
            end_date: End of date range

        Returns:
            DataFrame with all features indexed by as_of_date
        """
        logger.info(f"Computing crowd belief features for {specialist}")

        # Load raw beliefs
        df = self._load_beliefs(specialist, start_date, end_date)

        if df.empty:
            logger.warning(f"No crowd beliefs data for {specialist}")
            return pd.DataFrame()

        # Compute feature groups
        prob_features = self._compute_probability_features(df, specialist)
        momentum_features = self._compute_momentum_features(df, specialist)
        attention_features = self._compute_attention_features(df, specialist)
        consensus_features = self._compute_consensus_features(df, specialist)
        urgency_features = self._compute_urgency_features(df, specialist)

        # Merge all features
        result = prob_features

        for feature_df in [momentum_features, attention_features,
                          consensus_features, urgency_features]:
            if not feature_df.empty:
                result = result.merge(feature_df, on='as_of_date', how='left')

        logger.info(
            f"Generated {len(result.columns)-1} crowd belief features "
            f"for {specialist} across {len(result)} dates"
        )

        return result

    def compute_composite_signal(
        self,
        specialist: str,
        as_of_date: date
    ) -> Dict[str, float]:
        """
        Compute composite behavioral signal for a single date.

        Used for real-time inference and dashboard display.

        Returns:
            Dict with composite signals:
            - crowd_uncertainty: 0-1 (high = uncertain)
            - momentum_signal: -1 to +1 (directional)
            - attention_alert: bool (spike detected)
            - urgency_weight: 0-1 (high = near-term events)
        """
        # Fetch latest beliefs for this specialist
        df = self._load_beliefs(
            specialist,
            as_of_date - timedelta(days=1),
            as_of_date
        )

        if df.empty:
            return {
                'crowd_uncertainty': 0.5,
                'momentum_signal': 0.0,
                'attention_alert': False,
                'urgency_weight': 0.0,
            }

        latest = df[df['as_of_date'] == as_of_date]
        if latest.empty:
            latest = df.iloc[-1:]

        return {
            'crowd_uncertainty': 1 - latest['consensus_strength'].mean(),
            'momentum_signal': np.clip(
                latest['prob_momentum_7d'].mean() if 'prob_momentum_7d' in latest else 0,
                -1, 1
            ),
            'attention_alert': latest['attention_index_24h'].max() > 70,
            'urgency_weight': (1 / np.sqrt(
                latest['days_to_resolution'].min() or 365
            )) if 'days_to_resolution' in latest else 0.0,
        }


# =============================================================================
# CONVENIENCE FUNCTIONS
# =============================================================================

def get_crowd_features_for_training(
    conn: psycopg2.extensions.connection,
    specialist: str,
    start_date: date,
    end_date: date
) -> pd.DataFrame:
    """
    Get crowd belief features ready for specialist training.

    Convenience function that creates engine and computes features.

    Args:
        conn: Database connection
        specialist: Specialist bucket name
        start_date: Training start date
        end_date: Training end date

    Returns:
        DataFrame with features indexed by as_of_date
    """
    engine = CrowdBeliefsFeatureEngine(conn)
    return engine.compute_features(specialist, start_date, end_date)


def get_composite_crowd_signal(
    conn: psycopg2.extensions.connection,
    specialist: str,
    as_of_date: date
) -> Dict[str, float]:
    """
    Get composite crowd signal for real-time inference.

    Args:
        conn: Database connection
        specialist: Specialist bucket name
        as_of_date: Date for signal computation

    Returns:
        Dict with composite behavioral signals
    """
    engine = CrowdBeliefsFeatureEngine(conn)
    return engine.compute_composite_signal(specialist, as_of_date)


# =============================================================================
# INTEGRATION WITH TRUMP EFFECT SPECIALIST
# =============================================================================

def enhance_trump_effect_features(
    base_features: pd.DataFrame,
    conn: psycopg2.extensions.connection
) -> pd.DataFrame:
    """
    Enhance Trump Effect specialist features with crowd beliefs.

    This function integrates Polymarket signals into the existing
    TrumpEffectFeatureEngine output.

    Args:
        base_features: DataFrame from TrumpEffectFeatureEngine
        conn: Database connection

    Returns:
        Enhanced DataFrame with crowd belief features
    """
    if base_features.empty:
        return base_features

    start_date = base_features['as_of_date'].min()
    end_date = base_features['as_of_date'].max()

    # Get crowd features
    engine = CrowdBeliefsFeatureEngine(conn)
    crowd_features = engine.compute_features('trump_effect', start_date, end_date)

    if crowd_features.empty:
        logger.warning("No crowd belief data available for enhancement")
        return base_features

    # Merge on as_of_date
    enhanced = base_features.merge(
        crowd_features,
        on='as_of_date',
        how='left'
    )

    # Fill missing with neutral values
    crowd_cols = [c for c in enhanced.columns if c.startswith('trump_effect_')]
    for col in crowd_cols:
        if 'uncertainty' in col or 'consensus' in col:
            enhanced[col] = enhanced[col].fillna(0.5)
        elif 'momentum' in col or 'direction' in col:
            enhanced[col] = enhanced[col].fillna(0.0)
        else:
            enhanced[col] = enhanced[col].fillna(0.0)

    logger.info(
        f"Enhanced Trump Effect features with {len(crowd_cols)} crowd belief signals"
    )

    return enhanced


if __name__ == "__main__":
    # Test the feature engine
    import os
    from dotenv import load_dotenv

    load_dotenv()
    load_dotenv('.env.vercel')

    database_url = os.getenv("DATABASE_URL") or os.getenv("POSTGRES_URL")
    if not database_url:
        print("DATABASE_URL not found")
        exit(1)

    conn = psycopg2.connect(database_url)

    engine = CrowdBeliefsFeatureEngine(conn)
    features = engine.compute_features(
        specialist='trump_effect',
        start_date=date(2025, 1, 1),
        end_date=date.today()
    )

    print(f"\nComputed {len(features)} rows with columns:")
    print(features.columns.tolist())

    if not features.empty:
        print("\nSample data:")
        print(features.head())

    conn.close()

File: frontend/src/inngest/sources/markets/crowd-beliefs.ts
/**
 * ZINC-FUSION-V15: Polymarket Crowd Beliefs Ingestion
 *
 * Behavioral signal capturing crowd probability estimates for policy outcomes.
 * Forward-looking complement to backward-looking EPU indices.
 *
 * API: https://gamma-api.polymarket.com/events
 * Schedule: Daily at 6 PM ET (after market close)
 *
 * Cross-Specialist Routing:
 * - trump_effect: trump, executive, doge, deportation
 * - china: china, taiwan, trade_war
 * - tariff: tariff, import, trade
 * - biofuel: rfs, ethanol, epa, mandate
 * - energy: oil, sanctions, opec
 * - fed: fed, rates, inflation, recession
 * - volatility: vix, crash, crisis
 */

import { inngest } from "../../client";
import { prisma } from "@/lib/db";
import { Decimal } from "@prisma/client/runtime/library";

// =============================================================================
// TYPE DEFINITIONS
// =============================================================================

interface PolymarketMarket {
  id: string;
  question: string;
  outcomePrices: string[]; // [YES, NO] as string decimals
  liquidity: string;
  endDate: string;
}

interface PolymarketEvent {
  id: string;
  ticker: string;
  slug: string;
  title: string;
  volume: number;
  volume24hr: number;
  volume1wk: number;
  liquidity: number;
  active: boolean;
  closed: boolean;
  endDate: string;
  markets: PolymarketMarket[];
}

interface ProcessedBelief {
  eventSlug: string;
  outcomeQuestion: string;
  capturedAt: Date;
  impliedProbYes: number;
  impliedProbNo: number;
  attentionIndex24h: number;
  attentionIndex7d: number;
  probMomentum24h: number | null;
  probMomentum7d: number | null;
  consensusStrength: number;
  eventCategory: string;
  specialistTags: string[];
  eventResolutionDate: Date | null;
  daysToResolution: number | null;
  rawBettingVolumeUsd: number;
  rawLiquidityUsd: number;
}

// =============================================================================
// CATEGORY ROUTING
// =============================================================================

/**
 * Maps event categories to specialist tags for cross-specialist routing.
 * An event can route to multiple specialists.
 */
const CATEGORY_TO_SPECIALISTS: Record<string, string[]> = {
  tariff: ["tariff", "trump_effect", "china"],
  china: ["china", "trump_effect"],
  taiwan: ["china", "volatility"],
  trump: ["trump_effect"],
  doge: ["trump_effect", "fed"],
  executive: ["trump_effect"],
  deportation: ["trump_effect"],
  immigration: ["trump_effect"],
  fed: ["fed", "volatility"],
  rates: ["fed"],
  inflation: ["fed", "energy"],
  recession: ["fed", "volatility"],
  rfs: ["biofuel"],
  ethanol: ["biofuel", "energy"],
  biodiesel: ["biofuel"],
  epa: ["biofuel", "trump_effect"],
  mandate: ["biofuel"],
  oil: ["energy"],
  sanctions: ["energy", "china"],
  opec: ["energy"],
  drilling: ["energy", "trump_effect"],
  vix: ["volatility"],
  crash: ["volatility"],
  crisis: ["volatility"],
  trade: ["tariff", "china"],
  deficit: ["fed", "trump_effect"],
  spending: ["fed", "trump_effect"],
};

/**
 * Keywords to filter relevant events from Polymarket.
 * Only events containing these keywords are ingested.
 */
const RELEVANT_KEYWORDS = [
  "tariff",
  "china",
  "taiwan",
  "trump",
  "doge",
  "deport",
  "trade",
  "deficit",
  "spending",
  "fed",
  "rate",
  "inflation",
  "recession",
  "rfs",
  "ethanol",
  "biodiesel",
  "epa",
  "mandate",
  "oil",
  "sanction",
  "opec",
  "drill",
  "vix",
  "crash",
  "crisis",
  "immigration",
];

// =============================================================================
// HELPER FUNCTIONS
// =============================================================================

/**
 * Determines if an event is relevant for ZL forecasting.
 */
function isRelevantEvent(event: PolymarketEvent): boolean {
  const searchText = `${event.title} ${event.slug}`.toLowerCase();
  return RELEVANT_KEYWORDS.some((keyword) => searchText.includes(keyword));
}

/**
 * Extracts the primary category from event title/slug.
 */
function extractCategory(event: PolymarketEvent): string {
  const searchText = `${event.title} ${event.slug}`.toLowerCase();

  // Priority order matters - more specific first
  if (searchText.includes("tariff")) return "tariff";
  if (searchText.includes("taiwan")) return "taiwan";
  if (searchText.includes("china")) return "china";
  if (searchText.includes("doge")) return "doge";
  if (searchText.includes("deport") || searchText.includes("immigration"))
    return "deportation";
  if (searchText.includes("trump")) return "trump";
  if (searchText.includes("rfs") || searchText.includes("ethanol"))
    return "biofuel";
  if (searchText.includes("biodiesel") || searchText.includes("mandate"))
    return "biofuel";
  if (searchText.includes("oil") || searchText.includes("sanction"))
    return "energy";
  if (searchText.includes("fed") || searchText.includes("rate")) return "fed";
  if (
    searchText.includes("recession") ||
    searchText.includes("crash") ||
    searchText.includes("vix")
  )
    return "volatility";
  if (searchText.includes("deficit") || searchText.includes("spending"))
    return "deficit";
  if (searchText.includes("trade")) return "trade";

  return "other";
}

/**
 * Gets specialist tags for a category.
 */
function getSpecialistTags(category: string): string[] {
  return CATEGORY_TO_SPECIALISTS[category] || ["trump_effect"];
}

/**
 * Calculates consensus strength from YES/NO probabilities.
 * 1.0 = unanimous (one side at 100%)
 * 0.5 = maximally split (50/50)
 */
function calculateConsensusStrength(probYes: number): number {
  // Distance from 0.5 (maximum uncertainty)
  const distanceFrom50 = Math.abs(probYes - 0.5);
  // Scale to 0-1 where 1 = unanimous
  return distanceFrom50 * 2;
}

/**
 * Calculates attention index (normalized betting activity).
 * Returns 0-100 scale based on volume relative to baseline.
 */
function calculateAttentionIndex(
  volume: number,
  baselineVolume: number
): number {
  if (baselineVolume === 0) return 50; // Default baseline
  const ratio = volume / baselineVolume;
  // Cap at 100, scale logarithmically
  return Math.min(100, Math.log2(ratio + 1) * 50);
}

/**
 * Fetches prior beliefs for momentum calculation.
 */
async function fetchPriorBeliefs(
  eventSlug: string,
  outcomeQuestion: string,
  daysBack: number
): Promise<{ impliedProbYes: Decimal } | null> {
  const cutoffDate = new Date();
  cutoffDate.setDate(cutoffDate.getDate() - daysBack);

  const prior = await prisma.crowdBeliefsEvent.findFirst({
    where: {
      eventSlug,
      outcomeQuestion,
      capturedAt: { lte: cutoffDate },
    },
    orderBy: { capturedAt: "desc" },
    select: { impliedProbYes: true },
  });

  return prior;
}

// =============================================================================
// MAIN INGESTION FUNCTION
// =============================================================================

export const crowdBeliefs = inngest.createFunction(
  {
    id: "crowd-beliefs-daily",
    name: "Polymarket Crowd Beliefs",
    retries: 3,
  },
  { cron: "0 11 * * *" }, // 5AM CT = 11AM UTC, Daily
  async ({ event, step }) => {
    // Step 1: Fetch events from Polymarket Gamma API
    const events = await step.run("fetch-polymarket-events", async () => {
      const response = await fetch(
        "https://gamma-api.polymarket.com/events?active=true&closed=false",
        {
          headers: { Accept: "application/json" },
        }
      );

      if (!response.ok) {
        throw new Error(`Polymarket API error: ${response.status}`);
      }

      const data: PolymarketEvent[] = await response.json();
      return data;
    });

    // Step 2: Filter for relevant events
    const relevantEvents = await step.run("filter-relevant-events", async () => {
      return events.filter(isRelevantEvent);
    });

    console.log(
      `Found ${relevantEvents.length} relevant events out of ${events.length} total`
    );

    // Step 3: Process each event into beliefs
    const processedBeliefs = await step.run("process-beliefs", async () => {
      const beliefs: ProcessedBelief[] = [];
      const capturedAt = new Date();

      // Calculate baseline volume for attention index
      const totalVolume24h = events.reduce(
        (sum, e) => sum + (e.volume24hr || 0),
        0
      );
      const avgVolume24h = totalVolume24h / Math.max(events.length, 1);

      for (const event of relevantEvents) {
        const category = extractCategory(event);
        const specialistTags = getSpecialistTags(category);

        // Process each market (outcome) within the event
        for (const market of event.markets) {
          const probYes = parseFloat(market.outcomePrices[0] || "0");
          const probNo = parseFloat(market.outcomePrices[1] || "0");

          // Calculate attention index
          const attentionIndex24h = calculateAttentionIndex(
            event.volume24hr || 0,
            avgVolume24h
          );
          const attentionIndex7d = calculateAttentionIndex(
            event.volume1wk || 0,
            avgVolume24h * 7
          );

          // Calculate days to resolution
          let daysToResolution: number | null = null;
          let eventResolutionDate: Date | null = null;
          if (market.endDate) {
            eventResolutionDate = new Date(market.endDate);
            const diffMs = eventResolutionDate.getTime() - capturedAt.getTime();
            daysToResolution = Math.ceil(diffMs / (1000 * 60 * 60 * 24));
          }

          beliefs.push({
            eventSlug: event.slug,
            outcomeQuestion: market.question,
            capturedAt,
            impliedProbYes: probYes,
            impliedProbNo: probNo,
            attentionIndex24h,
            attentionIndex7d,
            probMomentum24h: null, // Will be calculated in next step
            probMomentum7d: null,
            consensusStrength: calculateConsensusStrength(probYes),
            eventCategory: category,
            specialistTags,
            eventResolutionDate,
            daysToResolution,
            rawBettingVolumeUsd: event.volume || 0,
            rawLiquidityUsd: event.liquidity || 0,
          });
        }
      }

      return beliefs;
    });

    // Step 4: Calculate momentum (requires DB lookup)
    const beliefsWithMomentum = await step.run(
      "calculate-momentum",
      async () => {
        const updatedBeliefs: ProcessedBelief[] = [];

        for (const belief of processedBeliefs) {
          // Lookup 24h prior
          const prior24h = await fetchPriorBeliefs(
            belief.eventSlug,
            belief.outcomeQuestion,
            1
          );

          // Lookup 7d prior
          const prior7d = await fetchPriorBeliefs(
            belief.eventSlug,
            belief.outcomeQuestion,
            7
          );

          updatedBeliefs.push({
            ...belief,
            probMomentum24h: prior24h
              ? belief.impliedProbYes - prior24h.impliedProbYes.toNumber()
              : null,
            probMomentum7d: prior7d
              ? belief.impliedProbYes - prior7d.impliedProbYes.toNumber()
              : null,
          });
        }

        return updatedBeliefs;
      }
    );

    // Step 5: Upsert to database
    const upsertResult = await step.run("upsert-beliefs", async () => {
      let inserted = 0;
      let updated = 0;

      for (const belief of beliefsWithMomentum) {
        try {
          await prisma.crowdBeliefsEvent.upsert({
            where: {
              eventSlug_outcomeQuestion_capturedAt: {
                eventSlug: belief.eventSlug,
                outcomeQuestion: belief.outcomeQuestion,
                capturedAt: belief.capturedAt,
              },
            },
            update: {
              impliedProbYes: belief.impliedProbYes,
              impliedProbNo: belief.impliedProbNo,
              attentionIndex24h: belief.attentionIndex24h,
              attentionIndex7d: belief.attentionIndex7d,
              probMomentum24h: belief.probMomentum24h,
              probMomentum7d: belief.probMomentum7d,
              consensusStrength: belief.consensusStrength,
              eventCategory: belief.eventCategory,
              specialistTags: belief.specialistTags,
              eventResolutionDate: belief.eventResolutionDate,
              daysToResolution: belief.daysToResolution,
              rawBettingVolumeUsd: belief.rawBettingVolumeUsd,
              rawLiquidityUsd: belief.rawLiquidityUsd,
            },
            create: {
              eventSlug: belief.eventSlug,
              outcomeQuestion: belief.outcomeQuestion,
              capturedAt: belief.capturedAt,
              impliedProbYes: belief.impliedProbYes,
              impliedProbNo: belief.impliedProbNo,
              attentionIndex24h: belief.attentionIndex24h,
              attentionIndex7d: belief.attentionIndex7d,
              probMomentum24h: belief.probMomentum24h,
              probMomentum7d: belief.probMomentum7d,
              consensusStrength: belief.consensusStrength,
              eventCategory: belief.eventCategory,
              specialistTags: belief.specialistTags,
              eventResolutionDate: belief.eventResolutionDate,
              daysToResolution: belief.daysToResolution,
              rawBettingVolumeUsd: belief.rawBettingVolumeUsd,
              rawLiquidityUsd: belief.rawLiquidityUsd,
              source: "polymarket",
            },
          });
          inserted++;
        } catch (error) {
          console.error(`Error upserting belief: ${belief.eventSlug}`, error);
        }
      }

      return { inserted, updated, total: beliefsWithMomentum.length };
    });

    return {
      success: true,
      eventsScanned: events.length,
      relevantEvents: relevantEvents.length,
      beliefsProcessed: upsertResult.total,
      timestamp: new Date().toISOString(),
    };
  }
);
